src/Conversion/ONNXToTOSA/Math/Conv2D.cpp

```diff
@@ -0,0 +1,122 @@
+ */
+
+//===---------------- Conv2D.cpp - Conv2D Op --------------------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX conv operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Utils/StaticValueUtils.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinAttributeInterfaces.h"
+#include "mlir/IR/BuiltinAttributes.h"
+#include "mlir/IR/BuiltinTypeInterfaces.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/TypeUtilities.h"
+#include "mlir/Support/LLVM.h"
+#include "mlir/Support/LogicalResult.h"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "llvm/ADT/None.h"
+#include "llvm/ADT/SmallVector.h"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXConvOpLoweringToTOSA : public OpConversionPattern<ONNXConvOp> {
+public:
+  using OpConversionPattern<ONNXConvOp>::OpConversionPattern;
+  using OpAdaptor = typename ONNXConvOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXConvOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto input = adaptor.X();
+    auto weights = adaptor.W();
+    auto bias = adaptor.B();
+
+    auto weightType = weights.getType().cast<ShapedType>();
+
+    auto weightShape = weightType.getShape();
+
+    StringAttr autopad = adaptor.auto_padAttr();
+    ArrayAttr dilations = adaptor.dilationsAttr();
+    IntegerAttr group = adaptor.groupAttr();
+    ArrayAttr pads = adaptor.padsAttr();
+    ArrayAttr strides = adaptor.stridesAttr();
+
+    Type resultType = getTypeConverter()->convertType(op.getResult().getType());
+
+    if (group.getSInt() != 1) {
+      return rewriter.notifyMatchFailure(
+          op, "ONNX Conv grouping not supported");
+    }
+
+    // NOTE: we would like if inferShapes() had filled in explicit padding
+    // but currently inferShapes() does not do this for ConvOp (it does for
+    // ConvTransposeOp). We have not implemented code for autopad so fail.
+    if (autopad && autopad != "NOTSET")
+      return op.emitError("padding must be explicit");
+
+    // Convert input [N,C,H,W] -> [N,H,W,C]
+    Value newInput =
+        tosa::createTosaTransposedTensor(rewriter, op, input, {0, 2, 3, 1});
+
+    // Convert weights [M,C,H,W] -> [M,H,W,C]
+    Value newWeight =
+        tosa::createTosaTransposedTensor(rewriter, op, weights, {0, 2, 3, 1});
+
+    if (bias.getType().isa<NoneType>()) {
+      DenseElementsAttr newBiasAttr = DenseElementsAttr::get(
+          RankedTensorType::get({weightShape[0]}, rewriter.getF32Type()),
+          {0.0F});
+      bias = rewriter.create<tosa::ConstOp>(
+          op->getLoc(), newBiasAttr.getType(), newBiasAttr);
```

Comment:
You can call tosa::getConstTensor here instead of manually creating the attribute and then the const operator

Reply:
Unfortunately, no. It does not accept single values with a shape that is other than 1

Reply:
Ah yes, the size is dynamic so you can't know in advance the shape, ok forget about it then

Reply:
It is actually not dynamic. It has a fixed size e.g., 32xf32, but all have value 0. The check `vec.size() != num_total_elements` does not allow something like thsis

Reply:
I will mark this as resolved :)